### PR TITLE
Initial tests and benchmark

### PR DESCRIPTION
**Pull request by pparke**
Since there was some performance profiling being done in the original tests I thought it might be good to formalize and record those benchmarks so we can easily see if code changes increase or decrease run time and also track improvements. I've added a benchmarks directory to the tests directory and added the `tinybench` package. You can run the benchmarks with `npm run bench` and when a code change is made that significantly improves the performance you can update the log file with `npm run bench:update` and then just commit the log file. The comparison uses the latest benchmark from the corresponding log file with a matching task name. If we need to add more than one task per benchmark then the code will need to be updated to support that. I've also used the Jest globals in the original test file to make those tests into Jest tests. I haven't changed how the tests themselves work other than using `expect` instead of throwing an error and changing the console log message. Hopefully that doesn't cause any conflicts but if it does they should be minor ones.

When you review, could you check the `tests/benchmarks/compareBenchmarks.ts` function in detail, specifically the way that I'm checking if the change in the mean execution time is significant. I'm not sure if my math is 100% correct or if I'm comparing the right values there. If there is a better way to detect a significant change then I'd be happy to update the code. The reference for available measurements that tinybench gives us is available here https://github.com/tinylibs/tinybench?tab=readme-ov-file#taskresult

Hope this is a helpful contribution!

### CODE REVIEW

**Comment by pparke** [comment]
Looks like I need to rebase!

**Comment by pparke** [comment]
Oh, I almost forgot to mention. I'm not sure if this is a bug or not but when you increase the number of iterations in the training loop for Autograd the loss will eventually become `NaN`. I didn't look into why that would happen but I added some error handling to report when it occurs and which iteration caused it.

**Comment by eduardoleao052** [comment]
@pparke this is amazing! It's definitely a great addition! I don't know what changes it is conflicting with, but I'm sure it won't be too bad to sort out. If you want any help, let me know.

About the NaN, it's true - that can happen in that training loop, I still need to investigate whether that's due to a bug or gradient overflow on an unstructured autograd (no Module). Thanks for pointing it out!

**Comment by eduardoleao052** [comment]
Maybe the testing is causing a conflict with the new test.ts file, because now there's a simple `jest` based test infrastructure, but I'm sure your benchmarks are a lot more precise. I'll review it later today. 
Thanks again!

**Comment by eduardoleao052** [comment]
Finally, I think you can PR it directly to main, I already merged the TS branch into it. Some new errors popped up, so I'm working on solving those as well.

**Comment by medic-code** [comment]
Nice addition, just a small comment, there's a typo in the readme addition `npm test` instead of `npm run test`!

Love the benchmarking, this was a thing i was thinking of when looking at the first iteration of the test prior to jest.

**Comment by medic-code** [comment]
Just another thought in the integration tests we are asserting the loss and quantifying performance.  

I wonder if this worth separating these two parts out and using benchmarking for quantifying performance.

**Comment by pparke** [comment]
@eduardoleao052 you're right, I made an error! It was meant to check by how much the new benchmark differed from the last +/- the margin of error. For example, if the new benchmark exceeded the last by 7% and the margin of error on the last was 4% then it was supposed to report that it exceeded the last margin of error by 3%. It looks like your code corrects that so I'll update it. Thanks!

**Comment by pparke** [comment]
On second thought, maybe I was just returning the wrong value but please let me know if it is meaningful to check if the current benchmark differed from the last by more than the last's margin of error. My stats knowledge is pretty rudimentary so I don't know if that's the best way to check for a significant change. I've updated the code so it returns the difference of the means as a percent but I kept the reporting so that it tells you how much it exceeded the last mean +/- margin of error.

Is that right? I'm assuming that the margin of error allows you to discard results that fall within it but maybe I'm missing something like should the margin of error of the new benchmark also be considered?

**Comment by pparke** [comment]
I've changed the base to `main` as well and I rewrote my changes to the `test.ts` file so that they keep your original test functions and then just use them in the `test` invocations. I added a `describe` block although I'm not sure if it's very useful at the moment since the individual tests are fairly independent. It's not necessary so I could remove it.

For building out more tests we should make sure to name the files as `[test suite name].spec.ts` or `[test suite name].test.ts` since that's what the jest config is looking for. Right now it's only working because the filename itself is `test` which is matching the regex. The tests you currently have are technically end to end tests or integration tests since they are checking if a bunch of functions work together to produce an end result. Future tests should include more integration tests that cover different ML applications as well as a few unit tests that check for possible bugs like if an incorrect configuration is provided when instantiating a class does the code correctly handle that by throwing a specific error https://jestjs.io/docs/expect#tothrowerror

**Comment by medic-code** [comment]
Yeah I've started adding some unit tests for utils, maybe a little over kill, called it utils.test.ts. In general, I agree with more integration tests though.

**Comment by eduardoleao052** [comment]
@pparke it is meaningful, because if the difference is smaller than the mean error of the last benchmark, that means that the difference is not significant, and is probably just random.
So it's important to check whether diffMean > lastBenchmark.rme before logging that there was a significant difference.

**Comment by eduardoleao052** [comment]
@pparke I'll quickly check the changes to the benchmark reporting right now, then I'll merge!

**Comment by eduardoleao052** [comment]
@pparke @medic-code to build and publish to `npm`, how does it work? I ran `npm run build`, published the package to `npm`, but only the `README.md`, the `LICENSE.md` and the `package.json` went to `npm`. How can I publish the whole package?

**Comment by eduardoleao052** [comment]
I think the "files" attribute points to "lib/**/*", should it not be "dist/*"?

**Comment by pparke** [comment]
That's my bad! Sorry I forgot to change that!
